### ISSUE TITLE
Droop in early, prompt and late correlation results 

### ISSUE THREAD

**Issue by rigoorozco** [noelpedro/WES_Capstone_GNSS_MR#5]
Hi @noelpedro,

I have a version of your HDL running on a ZCU102 with an fmcomms2/3 card. 

I'm seeing some strange behavior where my correlation results all drop and then re-adjust. This causes the receiver to miss a frame.

This missed frame creates a jump in time which then causes all sorts of issues and eventually make the receiver stop tracking.

Wanted to see if you had come across this yourself.

Thanks,

Rigo
![tracking-correlation-results-error](https://user-images.githubusercontent.com/12193068/208555046-0f656f7f-a412-41a3-bd99-2e596971cb07.png)

**Comment by noelpedro** [comment]
Hi, thanks for comment.

Wow you got it to work with fmcomms2?

Did you use a different tcx0?

Or using external reference?


So 23 seconds in it diverged.


The buffers can be overflowing you can monitor that with ILA.


Or maybe improvements needed in code resampler or carrier wipe off.

**Comment by rigoorozco** [comment]
Yeah, I removed the original XO, added a cap and now I’m using a signal generator that has good enough stability.

I think each sample here is 1 ms but could check. 

This seems something tricky for me to debug with ILAs since the FPGA is operating over many more samples than we see here—we’re only seeing the integrated results on the plots. 

I can check for overflows, I added a separate counter do check that or there was one already.

I’ve been staring at those to for some time now but I can’t seem to find what’s wrong.

**Comment by noelpedro** [comment]
Are u able to get PVT fix?

I noticed when the SW attempted to get a fix after tracking at least 4 birds and decoding the first 3 pages of the nav message. I would get fifo overflow because CPU usage would go up to 100%. Those overflows could happen as a result of PVT in SW.

Definitely seen overflows with 1ms integration period + PVT.

**Comment by rigoorozco** [comment]
I'm able to get a fix but it soon begins to accumulate error making it lose fix. 

I was playing around with gdb trying to figure out what's happening. What I think is happening is I'm getting a bad frame which causes a large increment in time in the tracking functions (not sure if I'm using the right terms here) which makes the velocity shoot up and everything goes crazy after.

On the ZCU102 I have 4 cores at 1.2 GHz so I actually don't see the CPU go up that high. Only seeing a load average about 0.6  while I'm getting position fixes. 

However, I don't seem to be getting any overflow counts when I do lose fix

**Comment by rigoorozco** [comment]
Okay, so I am getting some overflows prior to getting fix. But they seem to stop when it gets fix. 

CPU utilization seems pretty low still

**Comment by rigoorozco** [comment]
![image](https://user-images.githubusercontent.com/12193068/208777614-a503bc21-f1e1-4c3b-9d3e-7a3ef619dae0.png)
 
 I've been using 10 L1 C/A correlators

**Comment by noelpedro** [comment]
Ok so soon after the fix it crashes. So one of the channels must be overflowing. Are you monitoring the buffer overflows of each individual channel in the FPGA?

Also check the .LOG for PVT errors

**Comment by rigoorozco** [comment]
Yup each is counted individually.

I increased the buffer size from 32k to 64k. Don’t seem to be getting anymore overflows but still getting the same behavior. 

I’ve tried using the logging option but it seems to add more latency. I can try them again. Also need to dump tracking results again

**Comment by noelpedro** [comment]
Have you tried 20ms of integration?

**Comment by noelpedro** [comment]
You'll reduce interrupts from 1000 to 50 per second per channel during tracking if you do so. Not sure if you are at stationary location and clock is stable enough.

**Comment by pedromiguelcp** [comment]
Hi @rigoorozco & @noelpedro 

Sorry for interrupting this thread. Unfortunately, I can't help with this problem. So far I've only looked at the HW implementation of acquisition.

However, I have another question. What changes did you make to the gnss-sdr source code to allow offloading data to HW with the fmcomms2/3 card?
The ad9361_fpga_signal_source seems to support using HW accelerators (real-time & post-processing), however, fmcomms2_signal_source appears only to be compatible with the SW-only version of gnss-sdr. 
Did you simply compiled with DENABLE_FPGA & DENABLE_FMCOMMS2, and configure SignalSource.implementation=Fmcomms2_Signal_Source ?

Thanks in advance. Again, I apologize for writing in this thread.

**Comment by noelpedro** [comment]
It would be ad9361 signal source for fmcomms2/3 directly to hardware.

**Comment by rigoorozco** [comment]
@noelpedro my sim is stationary. Also, if connect my workstation and connect gnss-sdr to the dev board using iiod, it tracks just fine. So I don’t think it’s an RF or clock stability issue. 

I’ll increasing the correlation interval though

@pedromiguelcp I think there is also an input  dma (presumably for testing recordings) and an axi stream switch that the ad9361_fpga_signal_source driver requires

**Comment by noelpedro** [comment]
Using iiod and processing on the software doesn't generate 1000 interrupts per second to the cpu...

The cpu reaction time to the interrupts is undeterministic.

**Comment by rigoorozco** [comment]
For sure! I meant in regards to your point on location and clock stability. 

I’ll get back to testing in the new year. Thanks for the help, @noelpedro

**Comment by pedromiguelcp** [comment]
Hi @rigoorozco and @noelpedro, it's been a while. Have you made any progress on this? I'm working with a similar setup (ZCU102 + FMCOMMS5), but I am having some problems in the tracking stage, in particular, getting the signals to lock. When the tracking starts, I immediately have a drop in the correlator values.
Can you please share the gnss-sdr configuration file with me?
Also, could you let me know if you're using any filters to receive RF signals?
